pkg/tls/ocsp.go

```diff
@@ -0,0 +1,200 @@
+
+type ocspEntry struct {
+	leaf       *x509.Certificate
+	issuer     *x509.Certificate
+	responders []string
+	nextUpdate time.Time
+	staple     []byte
+}
+
+// ocspStapler retrieves staples from OCSP responders and store them in an in-memory cache.
+// It also updates the staples on a regular basis and before they expire.
+type ocspStapler struct {
+	client             *http.Client
+	cache              cache.Cache
+	forceStapleUpdates chan struct{}
+	responderOverrides map[string]string
+}
+
+// newOCSPStapler creates a new ocspStapler cache.
+func newOCSPStapler(responderOverrides map[string]string) *ocspStapler {
+	return &ocspStapler{
+		client:             &http.Client{Timeout: 10 * time.Second},
+		cache:              *cache.New(defaultCacheDuration, 5*time.Minute),
+		forceStapleUpdates: make(chan struct{}, 1),
+		responderOverrides: responderOverrides,
+	}
+}
+
+// Run updates the OCSP staples every hours.
+func (o *ocspStapler) Run(ctx context.Context) {
+	ticker := time.NewTicker(time.Hour)
+	defer ticker.Stop()
+
+	select {
+	case <-ctx.Done():
+		return
+
+	case <-o.forceStapleUpdates:
+		o.updateStaples(ctx)
+
+	case <-ticker.C:
+		o.updateStaples(ctx)
+	}
+}
+
+// ForceStapleUpdates triggers staple updates in the background instead of waiting for the Run routine to update them.
+func (o *ocspStapler) ForceStapleUpdates() {
+	select {
+	case o.forceStapleUpdates <- struct{}{}:
+	default:
+	}
+}
+
+// GetStaple retrieves the OCSP staple for the corresponding to the given key (public certificate hash).
+func (o *ocspStapler) GetStaple(key string) ([]byte, bool) {
+	if item, ok := o.cache.Get(key); ok && item != nil {
+		if entry, ok := item.(*ocspEntry); ok {
+			return entry.staple, true
+		}
+	}
+	return nil, false
+}
+
+// Upsert creates a new entry for the given certificate.
+// The ocspStapler will then be responsible from retrieving and updating the corresponding OCSP obtainStaple.
+func (o *ocspStapler) Upsert(key string, leaf, issuer *x509.Certificate) error {
+	if len(leaf.OCSPServer) == 0 {
+		return errors.New("leaf certificate does not contain an OCSP server")
+	}
+
+	if item, ok := o.cache.Get(key); ok {
+		o.cache.Set(key, item, cache.NoExpiration)
+		return nil
+	}
+
+	var responders []string
+	for _, url := range leaf.OCSPServer {
+		if len(o.responderOverrides) > 0 {
+			if newURL, ok := o.responderOverrides[url]; ok {
+				url = newURL
+			}
+		}
+		responders = append(responders, url)
+	}
+
+	o.cache.Set(key, &ocspEntry{
+		leaf:       leaf,
+		issuer:     issuer,
+		responders: responders,
+	}, cache.NoExpiration)
+
+	return nil
+}
+
+// ResetTTL resets the expiration time for all items that has no expiration.
+// This allows to set a TTL to cache that do not exist in the dynamic configuration anymore.
+// For those existing, the TTL will be set to zero when the Upsert method will be called during
+// the UpdateConfigs method of the TLS manager.
+func (o *ocspStapler) ResetTTL() {
```

Comment:
I'm wondering if setting a TTL for unused certificates is really worth it.
We'll be making OCSP calls every hour or on each `UpdateConfigs` run for certs that aren't even in use, and it adds extra complexity to the code.